File: Solutions/121. Best Time to Buy and Sell Stock/121. Best Time to Buy and Sell Stock.go
func maxProfit(p []int) int {
    res,minv:=0,math.MaxInt32
    for i,_:=range(p){
        if minv>p[i]{
            minv=p[i]
        }else{
            res=max(res,p[i]-minv)
        }
    }
    return res
}
func max(i,j int)int{
    if(i<j){
        return j
    }
    return i
}



//dp
func maxProfit(p []int) int {
    res,n:=0,len(p)
    record:=make([]int,n)
    for i:=1;i<n;i+=1{
        record[i]=max(0,record[i-1]+p[i]-p[i-1])
        res=max(res,record[i])
    }
    return res
}
func max(i,j int)int{
    if(i<j){
        return j
    }
    return i
}

File: Solutions/121. Best Time to Buy and Sell Stock/121. Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, p: List[int]) -> int:
        res,minv=0,sys.maxsize
        for i in range(len(p)):
            if minv>p[i]:
                minv=p[i]
            else:
                res=max(res,p[i]-minv)
        return res
        
        

#dp
class Solution:
    def maxProfit(self, p: List[int]) -> int:
        res,n=0,len(p)
        record=[0]*n
        for i in range(1,len(p)):
            record[i]=max(0,record[i-1]+p[i]-p[i-1])
            res=max(res,record[i])
        return res